Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -326,21 +327,78 @@ PhysicalParticleContainer::AddPlasmaFromFile(const std::string s_f, amrex::Real
     amrex::Real p_m = ps.second["mass"][openPMD::RecordComponent::SCALAR].loadChunk<amrex::Real>().get()[0];
     amrex::Real p_q = ps.second["charge"][openPMD::RecordComponent::SCALAR].loadChunk<amrex::Real>().get()[0];
-    int npart = ps.second["position"]["x"].getExtent()[0];
+    long npart = ps.second["position"]["x"].getExtent()[0];
     series.flush();
 
+    //Conversion from Geant4 system of units (http://geant4.web.cern.ch/sites/geant4.web.cern.ch/files/geant4/collaboration/working_groups/electromagnetic/gallery/units/SystemOfUnits.html) to WarpX (SI)
     mass = p_m*PhysConst::mevpc2_kg;
     charge = p_q*PhysConst::q_e;
-    Real const weight = q_tot/(charge*amrex::Real(npart));
+    amrex::Real mmpns_mps = 1.e6;
+    amrex::Real mm_m = 1.e-3;
+    amrex::Real weight;
+    if (physical_q_tot!=0.0){
+        weight = physical_q_tot/(charge*amrex::Real(npart));
+    }
+    else {
+        //In case openPMD file contains macroparticle instead of real particle data
+        weight = charge;
+    }
+
+    std::shared_ptr<amrex::Real> ptr_x = ps.second["position"]["x"].loadChunk<amrex::Real>();
+    series.flush();
+    std::shared_ptr<amrex::Real> ptr_vx = ps.second["velocity"]["x"].loadChunk<amrex::Real>();
+    series.flush();
+    std::shared_ptr<amrex::Real> ptr_z = ps.second["position"]["z"].loadChunk<amrex::Real>();
+    series.flush();
+    std::shared_ptr<amrex::Real> ptr_vz = ps.second["velocity"]["z"].loadChunk<amrex::Real>();
+    series.flush();
+#if (defined WARPX_DIM_3D)
+    std::shared_ptr<amrex::Real> ptr_y = ps.second["position"]["y"].loadChunk<amrex::Real>();
+    series.flush();
+    std::shared_ptr<amrex::Real> ptr_vy = ps.second["velocity"]["y"].loadChunk<amrex::Real>();
+    series.flush();
+#endif
 
-    amrex::Print() << npart << " parts of species " << ps.first << "\nWith"
-    << " mass = " << mass << " and charge = " << charge << "\nTo initialize "
-    << npart << " macroparticles with weight " << weight << "\n";
+    // Allocate temporary vectors on the CPU
+    Gpu::HostVector<ParticleReal> particle_w;
+    Gpu::HostVector<ParticleReal> particle_x;
+    Gpu::HostVector<ParticleReal> particle_y;
+    Gpu::HostVector<ParticleReal> particle_z;
+    Gpu::HostVector<ParticleReal> particle_ux;
+    Gpu::HostVector<ParticleReal> particle_uy;
+    Gpu::HostVector<ParticleReal> particle_uz;
 
-    amrex::Print()<<"WARNING: this is WIP, no particle has been injected!!";
+    if (ParallelDescriptor::IOProcessor()) {
+        for (long i = 0; i < npart; ++i) {
+            amrex::Real x = ptr_x.get()[i]*mm_m;
+            amrex::Real ux = ptr_vx.get()[i]*mmpns_mps;
+            amrex::Real z = ptr_z.get()[i]*mm_m;
+            amrex::Real uz = ptr_vz.get()[i]*mmpns_mps;
+#if (defined WARPX_DIM_3D)
+            amrex::Real y = ptr_y.get()[i]*mm_m;
+            amrex::Real uy = ptr_vy.get()[i]*mmpns_mps;
+#elif (defined WARPX_DIM_XZ)
+            amrex::Real y = 0.0;
+            amrex::Real uy = 0.0;
```

Comment:
The are all `amrex::ParticleReal const` types :)

Reply:
Oh sorry, I remembered to check for charge and mass to be const, but I forgot to change these. Will do that now. Thanks for catching this @ax3l !

Reply:
(and `ParticleReal` instead of `Real` :) )

Reply:
This is done and tested in the next commit!